src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,374 @@
+    'mexh',
+    'morl',
+    'cgau1', 'cgau2', 'cgau3', 'cgau4', 'cgau5', 'cgau6', 'cgau7', 'cgau8',
+    'shan',
+    'fbsp',
+    'cmor'
+]
+class WaveletOp(LinearOperator):
+    """Wavelet operator class."""
+
+    def __init__(
+        self,
+        domain_shape: Sequence[int] | None = None,
+        dim: tuple[int] | tuple[int, int] | tuple[int, int, int] = (-2, -1),
+        wavelet_name: WaveletType = 'db4',
+        level: int | None = None,
+    ):
+        """Wavelet operator.
+
+        For complex images the wavelet coefficients are calculated for real and imaginary part separately.
+
+        For a 2D image, the coefficients are labelled [aa, (ad_n, da_n, dd_n), ..., (ad_1, da_1, dd_1)] where a refers
+        to the approximation coefficients and d to the detail coefficients. The index indicates the level.
+
+        Parameters
+        ----------
+        domain_shape
+            Shape of domain where wavelets are calculated. If set to None the shape is taken from the input of the
+            forward operator. The adjoint operator will raise an error.
+        dim
+            Dimensions (axes) where wavelets are calculated
+        wavelet_name
+            Name of wavelets
+        level
+            Highest wavelet level. If set to None, the highest possible level is calculated based on the domain shape.
+
+        Raises
+        ------
+        ValueError
+            If wavelets are calculated for more than three dimensions.
+        ValueError
+            If wavelet dimensions and domain shape do not match.
+        NotImplementedError
+            If any dimension of the domain shape is odd. Adjoint will lead to the wrong domain shape.
+        """
+        super().__init__()
+        self._domain_shape = domain_shape
+        self._wavelet_name = wavelet_name
+        self._dim = dim
+        self._level = level
+
+        # number of wavelet directions
+        if len(dim) == 1:
+            self.n_wavelet_directions = 1
+        elif len(dim) == 2:
+            self.n_wavelet_directions = 3
+        elif len(dim) == 3:
+            self.n_wavelet_directions = 7
+        else:
+            raise ValueError('Only 1D, 2D and 3D wavelet transforms are supported.')
+
+        if domain_shape is not None:
+            if len(dim) != len(domain_shape):
+                raise ValueError(
+                    'Number of dimensions along which the wavelet transform should be calculated needs to',
+                    'be the same as the domain shape',
+                )
+
+            if any(d % 2 for d in domain_shape):
+                raise NotImplementedError(
+                    'ptwt only supports wavelet transforms for tensors with even number of '
+                    'entries for all considered dimensions.'
+                )
+
+            # size of wavelets
+            wavelet_length = torch.as_tensor((Wavelet(wavelet_name).dec_len,) * len(domain_shape))
```

Comment:
Optimize the calculation of `wavelet_length` and `current_shape` by precomputing these values if possible, especially if they do not depend on dynamic input data. This could reduce computational overhead. [medium]